### ISSUE TITLE
HasOne Relationships missing targetName

### ISSUE THREAD

**Issue by lawmicha** [aws-amplify/amplify-ios#920]
When I provision my persistance layer with the following schema, and perform operations on the generated Models, I am unable to successfully persist and retrieve the data.

The schema used to provision the SQL database and the AppSync API:
```
type Project2 @model {
  id: ID!
  name: String
  teamID: ID! // Explicit field for Team association
  team: Team2 @connection(fields: ["teamID"])
}

type Team2 @model {
  id: ID!
  name: String!
}
```
Normally, when creating a connection without an explicit field for the connection, the provisioned API provides a create mutation API that takes in an input field called "projectTeamID" for you to associate a team with the project when performing create mutation with project.

However, because the create mutation API takes in 'teamID" instead of the `projectTeamId` as the field for the association. The GraphQL request that needs to be generated is:
```
mutation CreateProject {
  createProject2(input: {teamID: "teamId123"}) {
    createdAt
    id
    name
    teamID
    updatedAt
  }
}
```

The generated model is
```swift
public struct Project2: Model {
  public let id: String
  public var name: String?
  public var teamID: String
  public var team: Team2?
  
  public init(id: String = UUID().uuidString,
      name: String? = nil,
      teamID: String,
      team: Team2? = nil) {
      self.id = id
      self.name = name
      self.teamID = teamID
      self.team = team
  }
}
extension Project2 {
  // MARK: - CodingKeys 
   public enum CodingKeys: String, ModelKey {
    case id
    case name
    case teamID
    case team
  }
  
  public static let keys = CodingKeys.self
  //  MARK: - ModelSchema 
  
  public static let schema = defineSchema { model in
    let project2 = Project2.keys
    
    model.pluralName = "Project2s"
    
    model.fields(
      .id(),
      .field(project2.name, is: .optional, ofType: .string),
      .field(project2.teamID, is: .required, ofType: .string),
      .hasOne(project2.team, is: .optional, ofType: Team2.self, associatedWith: Team2.keys.id)
    )
    }
}
public struct Team2: Model {
  public let id: String
  public var name: String
  
  public init(id: String = UUID().uuidString,
      name: String) {
      self.id = id
      self.name = name
  }
}
extension Team2 {
  // MARK: - CodingKeys 
   public enum CodingKeys: String, ModelKey {
    case id
    case name
  }
  
  public static let keys = CodingKeys.self
  //  MARK: - ModelSchema 
  
  public static let schema = defineSchema { model in
    let team2 = Team2.keys
    
    model.pluralName = "Team2s"
    
    model.fields(
      .id(),
      .field(team2.name, is: .required, ofType: .string)
    )
    }
}

```

The call fails when I try to create the data using Amplify.API

```
// First create the team, and persist it
let team2 = Team2(name: "teamName")
Amplify.API.mutate(request: .create(team2))
// Associate the team in your project by passing in as the team object
let project2 = Project2(team: team2)
Amplify.API.mutate(request: .create(project2)) // It should be generating GraphQL request with `teamId == team2.id`  but does not
```

This is broken for API because the GraphQL input translated a:
```
{
 "teamID": null, // "teamID" field not used so it is empty
  "team": "teamName" // team object translated to "team": team2.id
}
```

The codegen fix is to provide the targetName as part of the `hasOne` relationship and remove the explicit `teamID` field from Project2
```swift
public struct Project2: Model {
  public let id: String
  public var name: String?
  public var team: Team2?
  // ...
}
extension Project2 {
   // ...
    model.fields(
        .hasOne(project2.team, is: .optional, ofType: Team2.self, associatedWith: Team2.keys.id, targetName: "teamID")
    )
}
```

With this, the GraphQL request will translate the team object over as `teamID == team2.id`

**Comment by lawmicha** [comment]
I believe this is related to https://github.com/aws-amplify/amplify-ios/issues/880 due to the use of both `doctorId` and passing `doctor` object into the Case Model

**Comment by lawmicha** [comment]
For a non-breaking change for developers, we will continue to support the two fields that are generated. the proposal for codegen is just passing the targetName to the hasOne relationship
```swift
public struct Project2: Model {
  public let id: String
  public var name: String?
  public var teamID: String
  public var team: Team2?
  // ...
}
extension Project2 {
   // ...
    model.fields(
        .hasOne(project2.team, is: .optional, ofType: Team2.self, associatedWith: Team2.keys.id, targetName: "teamID")
    )
}
```

**Comment by lawmicha** [comment]
## Current state
### Making the call with Amplify.API
Save the team first, then take the team and save the Project with the team associated
```swift
let project = Project( /scenario/ )
Amplify.API.mutate(request: .create(project))
```
1. teamID only
`let project = Project(teamID: team.id)`
Generated GraphQL input
```
{
   teamId: team.id
   team: null -> fails at runtime due to "team" is not a field on the API, potential bug fix that removes `team` key from the input
}
```

2. team object only - Not a case- teamID is required

3. teamID and team object
`let project = Project(teamID: team.id, team: team)`
```
{
teamId: team.id -> 
team: team.id  -> fails at runtime due to "team" is not a field on the API, potential bug fix that removes `team` key from the input
}
```

### Making the call with Amplify.DataStore 
sync to cloud will fail since all API calls above will fail
```swift
let project = Project( /scenario/ )
Amplify.DataStore.save(project)
```
1. teamID only
`let project = Project(teamID: team.id)`

DataStore SQL 
```
?
```
2. team object only - Not a case- teamID is required

3. teamID and team object
`let project = Project(teamID: team.id, team: team)`

DataStore SQL
```
?
```

**Comment by palpatim** [comment]
Released in [1.5.1](https://github.com/aws-amplify/amplify-ios/releases/tag/v1.5.1)
